### PR TITLE
Improve Dispatch Errors

### PR DESCRIPTION
**Pull request by jsdw**
Dispatch errors are a bit pants; we only decode the Module variant, and then we cloned a bunch of details out of metadata into it. This PR makes the following changes:

- Expose more dispatch error details; this is important since token errors are now separate from module errors, and are quite a likely thing people might want to react to.
- Use the `DispatchError` in our `dry_run` result too, so we get the same level of detail there (for dispatch erros at least).
- Re-work `ModuleError` API; don't pull/clone things from metadata; call that on-demand instead.
- Use `DecodeAsType` to decode `DispatchErrors` based on shape, so we don't have to worry about indexes changing if variants are added.
- Mark all errors `#[non_exhaustive]` to future proof a little against new variants being added (with `DecodeAsType`, this at least gives us a path to add new variants and deprecate old ones without worrying about index stuff).

### CODE REVIEW

File: subxt/src/error/mod.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    SubscriptionDropped,
+}
+
+/// Block error
+#[derive(Clone, Debug, Eq, thiserror::Error, PartialEq)]
+#[non_exhaustive]
+pub enum BlockError {
+    /// The block
+    #[error("Could not find a block with hash {0} (perhaps it was on a non-finalized fork?)")]
+    BlockHashNotFound(String),
```
Comment (niklasad1) [inline subxt/src/error/mod.rs:110]: I think it's sufficient to have BlockError, no need to duplicate the enum variant as well

```suggestion
    HashNotFound(String),
```

File: subxt/src/error/mod.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    StorageAddress(#[from] StorageAddressError),
+    /// The bytes representing an error that we were unable to decode.
+    #[error("An error occurred but it could not be decoded: {0:?}")]
+    Unknown(Vec<u8>),
+    /// Other error.
+    #[error("Other error: {0}")]
+    Other(String),
+}
+
+impl From<&str> for Error {
```
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/error/mod.rs:71]: But impl `From<T>` I think would conflict with all of the other `From<InnerError>` impls we have on `Error` :(

Let me double check that..

[the remark this answers is not in the document]
Comment (niklasad1) [reply subxt/src/error/mod.rs:71]: Ok, I did exactly the same for a while ago and it inferred it as `'static` maybe just me then :)

File: subxt/src/error/mod.rs (hunk 1 above)
Comment (jsdw) [reply subxt/src/error/mod.rs]: Good point; I prefer that; thanks!

**Review by niklasad1 — APPROVED** [review]
:+1:

File: subxt/src/error/mod.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+#[non_exhaustive]
+pub enum TransactionError {
+    /// The finality subscription expired (after ~512 blocks we give up if the
+    /// block hasn't yet been finalized).
+    #[error("The finality subscription expired")]
+    FinalitySubscriptionTimeout,
+    /// The block hash that the transaction was added to could not be found.
+    /// This is probably because the block was retracted before being finalized.
+    #[error("The block containing the transaction can no longer be found (perhaps it was on a non-finalized fork?)")]
+    BlockHashNotFound,
```
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/error/mod.rs]: And this too!

File: subxt/src/error/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsdw) [reply subxt/src/error/mod.rs]: I put an explicit lifetime on this to make it clear, and because I'm surprised I didn't need one already!

**Review by lexnv — APPROVED** [review]
LGTM!

**Pull request merged**